src/top/run/box_generator.py

```diff
@@ -0,0 +1,227 @@
+                begin, end = EDGES[axis * 4 + edge_id]
+                scale[axis] += np.linalg.norm(vertices[begin, :] - vertices[end, :])
+            scale[axis] /= 4.
+
+        x = cls.scaled_axis_aligned_vertices(scale)
+        system = np.concatenate((x, np.ones((NUM_KEYPOINTS, 1))), axis=1)
+        solution, _, _, _ = lstsq(system, vertices, rcond=None)
+        orientation = solution[:3, :3].T
+        translation = solution[3, :3]
+        
+        return orientation, translation, scale
+
+    def inside(self, point):
+        """
+        Tests whether a given point is inside the box.
+        Brings the 3D point into the local coordinate of the box.
+        In the local coordinate, the looks like an axis-aligned bounding box.
+        Next checks if the box contains the point.
+        Args:
+            point: A 3*1 numpy vector.
+        Returns:
+            True if the point is inside the box, False otherwise.
+        """
+        inv_trans = np.linalg.inv(self.transformation)
+        scale = self.scale
+        point_w = np.matmul(inv_trans[:3, :3], point) + inv_trans[:3, 3]
+        for i in range(3):
+            if abs(point_w[i]) > scale[i] / 2.:
+                return False
+        return True
+
+    def sample(self):
+        """Samples a 3D point uniformly inside this box."""
+        point = np.random.uniform(-0.5, 0.5, 3) * self.scale
+        point = np.matmul(self.rotation, point) + self.translation
+        return point
+
+    @property
+    def vertices(self):
+        return self._vertices
+
+    @property
+    def rotation(self):
+        if self._rotation is None:
+            self._rotation, self._translation, self._scale = self.fit(self._vertices)
+        return self._rotation
+
+    @property
+    def translation(self):
+        if self._translation is None:
+            self._rotation, self._translation, self._scale = self.fit(self._vertices)
+        return self._translation
+
+    @property
+    def scale(self):
+        if self._scale is None:
+            self._rotation, self._translation, self._scale = self.fit(self._vertices)
+        return self._scale
+
+    @property
+    def volume(self):
+        """
+        Compute the volume of the parallelpiped or the box.
+        For the boxes, this is equivalent to np.prod(self.scale).
+        However for parallelpiped, this is more involved.
+        Viewing the box as a linear function we can estimate the volume using a determinant.
+        This is equivalent to sp.ConvexHull(self._vertices).volume
+        Returns:
+            volume (float)
+        """
+        if self._volume is None:
+            i = self._vertices[2, :] - self._vertices[1, :]
+            j = self._vertices[3, :] - self._vertices[1, :]
+            k = self._vertices[5, :] - self._vertices[1, :]
+            sys = np.array([i, j, k])
+            self._volume = abs(np.linalg.det(sys))
+        return self._volume
+
+    @property
+    def transformation(self):
```

Comment:
Keeping around both `_transformation` and `_rotation,_translation` feels like a recipe for error (inconsistency) to me. One of them should be a derived property of another.

Reply:
We could instead prefer:
```
@property
def rotation(self):
    # ( check if _transformation exists here, if you'd like )
    return self._transformation[:3,:3]
```